### ISSUE TITLE
[models]: models.common SamplingGenerator does not guard blockfloats before applying penalties

### ISSUE THREAD

**Issue by tchedaTT** [tenstorrent/tt-metal#47002]
### Component / Area

models

### Issue Type (optional)

Bad Outputs

### Observed

The SamplingGenerator class, which handles penalties application and then the actual sampling, does not have a guard against blockfloat inputs.
tt_sampling.py does typecast to bf16, but that is after the penalties have been applied.
This means operations that were meant to be elementwise can leak across tokens as some logits take extreme values.
llama3_70b_galaxy and tt transformers seem to use bpf8_b logits at this point, so they are potentially affected.
Fortunately, it appears that it cannot leak across users (unless potentially running with transpose_tile=True). The blockfloats share the exponent among 16 numbers in one row, and entries in a row are different tokens of same user.

### Expected

cast to bf16 before passing to the module. Guard for bf16 in module.

### 1. Steps (exact commands)

-

### 2. Input data / link or description

-

### 3. Frequency

-

### 1. Software Versions

-

### 2. Hardware Details

-

### Is this a regression?

No

### Priority

None
